source/extensions/filters/http/grpc_web/grpc_web_filter.cc

```diff
@@ -156,14 +175,44 @@ Http::FilterHeadersStatus GrpcWebFilter::encodeHeaders(Http::ResponseHeaderMap&
   } else {
     headers.setReferenceContentType(Http::Headers::get().ContentTypeValues.GrpcWebProto);
   }
-  return Http::FilterHeadersStatus::Continue;
+
+  if (end_stream || isValidResponseHeaders(headers, end_stream)) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  ENVOY_LOG(debug, "received invalid response headers since the upstream response or local reply "
+                   "is not a gRPC or gRPC-Web response");
+  response_headers_ = &headers;
+  return Http::FilterHeadersStatus::StopIteration;
 }
 
-Http::FilterDataStatus GrpcWebFilter::encodeData(Buffer::Instance& data, bool) {
+Http::FilterDataStatus GrpcWebFilter::encodeData(Buffer::Instance& data, bool end_stream) {
   if (!is_grpc_web_request_) {
     return Http::FilterDataStatus::Continue;
   }
 
+  // When the upstream response (this is also relevant for local reply, since gRPC-Web request is
+  // not a gRPC request which makes the local reply's is_grpc_request set to false) is not a gRPC
+  // response, we set the "grpc-message" header with the upstream body content.
+  if (response_headers_ != nullptr) {
+    if (!end_stream) {
+      return Http::FilterDataStatus::StopIterationAndBuffer;
```

Comment:
It is possible that upstream send large response and then we buffer a lot here?

Reply:
Are you suggesting `StopIterationAndNoBuffer`? I have it there but since https://github.com/envoyproxy/envoy/pull/14516#discussion_r552968560. I think I misunderstood the statement 😅. will revert back.

Reply:
Not really, you can buffer until 1024 bytes then throw the rest.